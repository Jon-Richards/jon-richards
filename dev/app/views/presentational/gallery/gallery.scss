@import '~scss-standards';

.root {
  display: flex;
  width: 100%;
  height: auto;
  flex-wrap: wrap;
  
  @media (min-width: std-get-breakpoint(480)) {
    align-items: center;
  }
}
