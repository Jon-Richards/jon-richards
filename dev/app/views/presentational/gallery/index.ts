export { Gallery, GalleryProps } from './gallery';
