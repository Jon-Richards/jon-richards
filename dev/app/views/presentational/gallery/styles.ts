import { Breakpoints } from 'Config/styles/breakpoints';
import { Dimensions } from 'Config/styles/dimensions';

/** Styles for the Gallery component. */
export const STYLES = {
  root: `
    display: grid;
    width: 100%;
    height: auto;
    margin-bottom: ${Dimensions.whitespace};
    grid-template-columns: 1fr;
    grid-template-rows: repeat(auto-fill, auto);
    column-gap: ${Dimensions.whitespace};
    row-gap: ${Dimensions.whitespace};

    @media (min-width: ${Breakpoints[720]}) {
      grid-template-columns: 1fr 1fr;
    }

    @media (min-width: ${Breakpoints[1440]}) {
      grid-template-columns: 1fr 1fr 1fr;
    }
  `,
};
