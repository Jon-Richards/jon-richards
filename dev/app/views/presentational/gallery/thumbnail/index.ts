export { Thumbnail, ThumbnailProps } from './thumbnail';
