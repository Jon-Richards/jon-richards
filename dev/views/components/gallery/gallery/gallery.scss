/**
 * @fileoverview
 * Styling for the gallery component.
 */

@import '../mediator';

.root {
  display: flex;
  position: relative;
  width: 100%;
  height: 100%;
  flex-direction: column;
}
