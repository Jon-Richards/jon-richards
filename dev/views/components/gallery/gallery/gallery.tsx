/**
 * @fileoverview
 * 
 * Contains the gallery component.
 */

const CSS = require('./gallery.scss');
import { 
    React, 
    Piece,
    PieceShape
} from '../mediator';

/** Renders a gallery of portfolio pieces. */
export class Gallery extends React.Component<
    {
        /** Array of thumnail information for portfolio pieces. */
        pieces: PieceShape[];
    },
    never
> {
    /** Maps piece to Piece components. */
    private mapPieces = (): React.ReactFragment[] => {
        const pieces: React.ReactFragment[] = this.props.pieces.map( piece => {
            return (
                <Piece
                    key={piece.uuid}
                    uuid={piece.uuid}
                    title={piece.title}
                    description={piece.description}
                />
            );
        });
        return pieces;
    }

    /** Renders the gallery to the DOM. */
    render(): JSX.Element {
        return (
            <div className={CSS['root']}>
                {this.mapPieces()}
            </div>
        );
    }
}