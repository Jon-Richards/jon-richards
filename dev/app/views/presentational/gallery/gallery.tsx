/** @jsx jsx */

import * as React from 'react';
import { css, jsx} from '@emotion/core';
import { Thumbnail, ThumbnailProps } from './thumbnail';
import { STYLES } from './styles';

type ThumbnailShape = {
  /** A unique identifier for this thumbnail. */
  uuid: string;
  /** The image's source url for small screens. (phones) */
  sourceSmall: string;
  /** The image's source url for medium screens. (tablets, low-end laptops) */
  sourceMedium: string;
  /** The image's source url for large screens. (freestanding monitors) */
  sourceLarge: string;
  /** A very short description of the project that the thumbnail represents. */
  description: ThumbnailProps['altText'];
  /** The destination for the thumbnail's link. */
  href: ThumbnailProps['href'];
};

/** For computing the size of the image that will be loaded by a thumbnail. */
type ThumbnailSize = 'SMALL' | 'MEDIUM' | 'LARGE';

/** Props used by the Gallery component. */
export interface GalleryProps {
  /**
   * Information about the thumbnails that should be displayed in the gallery.
   */
  thumbnails: ThumbnailShape[];
  /** The size at which thumbnails should render. */
  thumbnailSize: ThumbnailSize;
  /** Click handler that passes the URI of the thumbnail that was clicked. */
  onClick: (path: string) => {};
}

function computeThumbnailSource (
  size: ThumbnailSize,
  thumbnail: ThumbnailShape
): string {
  switch(size) {
    case 'LARGE':
      return thumbnail.sourceLarge;
    case 'MEDIUM':
      return thumbnail.sourceMedium;
    default:
      return thumbnail.sourceSmall;
  }
}

function mapThumbnails(
  thumbnails: ThumbnailShape[],
  size: ThumbnailSize,
  onClick: GalleryProps['onClick']
): JSX.Element {
  const mapped = thumbnails.map(thumbnail => {
    const {
      description,
      uuid,
      href
    } = thumbnail;
    
    const src = computeThumbnailSource(size, thumbnail);

    return(
      <Thumbnail
        key={uuid}
        src={src}
        altText={description}
        href={href}
        onClick={() => onClick(href)}
      />
    );
  });

  return <React.Fragment>{mapped}</React.Fragment>;
}

function Gallery (props: GalleryProps): JSX.Element {
  const { thumbnails, thumbnailSize, onClick } = props;
  const mappedThumbs = mapThumbnails(thumbnails, thumbnailSize, onClick);

  return (<div css={css(STYLES.root)}>{mappedThumbs}{mappedThumbs}</div>);
}

const GALLERY_MEMO = React.memo<GalleryProps>(Gallery);

export { GALLERY_MEMO as Gallery };