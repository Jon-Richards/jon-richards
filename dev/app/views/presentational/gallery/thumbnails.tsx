import * as React from 'react';
import { Thumbnail, ThumbnailProps } from './thumbnail';
import { v4 as uuid } from 'uuid';

export function Thumbnails(props: ThumbnailsProps): JSX.Element {
  const mapped = props.data.map(thumbnail => {
    const { sources, fallbackSource, altText, href } = thumbnail;

    return (
      <Thumbnail
        key={uuid()}
        sources={sources}
        fallbackSource={fallbackSource}
        altText={altText}
        href={href}
        onClick={e => props.clickHandler(e, href)}
      />
    );
  });

  return <React.Fragment>{mapped}</React.Fragment>;
}

export type ThumbnailsProps = {
  data: ThumbnailShape[],
  clickHandler: (
    e: React.MouseEvent<HTMLElement, MouseEvent>,
    href: string
  ) => {}
}

type ThumbnailShape = Omit<ThumbnailProps[][0], 'onClick'>;
