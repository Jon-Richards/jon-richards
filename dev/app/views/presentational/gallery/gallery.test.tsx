import * as React from 'react';
import { render } from 'enzyme';
import { Gallery, GalleryProps } from './gallery';

describe('The gallery component.', () => {
  it('Should match the snapshot.', () => {
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const stubHandler = function () {};
    const test_1 = render(
      <Gallery
        thumbnails={[]}
        thumbnailSize="MEDIUM"
        onClick={() => stubHandler}
      />
    );

    expect(test_1).toMatchSnapshot();
  });

  it('Should render the one thumbnail for each in the array.', () => {
    const stubHandler = function () {};
    const thumbs: GalleryProps['thumbnails'] = [
      {
        uuid: 'abc-123',
        sourceSmall: 'small.jpg',
        sourceMedium: 'medium.jpg',
        sourceLarge: 'large.jpg',
        description: 'a thumbnail',
        href: 'www.google.com'
      },
      {
        uuid: 'abc-456',
        sourceSmall: 'small.jpg',
        sourceMedium: 'medium.jpg',
        sourceLarge: 'large.jpg',
        description: 'a thumbnail',
        href: 'www.google.com'
      }
    ];
    const test_1 = render(
      <Gallery
        thumbnails={thumbs}
        thumbnailSize="MEDIUM"
        onClick={() => stubHandler}
      />
    );

    expect(test_1.find('[data-testid="gallery__thumb"]').length).toBe(2);
  });

  it('Should apply the correct image source based on the size.', () => {
    const stubHandler = function () {};
    const thumbs: GalleryProps['thumbnails'] = [
      {
        uuid: 'abc-123',
        sourceSmall: 'small.jpg',
        sourceMedium: 'medium.jpg',
        sourceLarge: 'large.jpg',
        description: 'a thumbnail',
        href: 'www.google.com'
      },
      {
        uuid: 'abc-456',
        sourceSmall: 'small.jpg',
        sourceMedium: 'medium.jpg',
        sourceLarge: 'large.jpg',
        description: 'a thumbnail',
        href: 'www.google.com'
      }
    ];
    const test_1 = render(
      <Gallery
        thumbnails={thumbs}
        thumbnailSize="MEDIUM"
        onClick={() => stubHandler}
      />
    );

    expect(test_1.find('[data-testid="gallery__thumb__img"]')
      .attr('src'))
      .toBe('medium.jpg');
  });
});
