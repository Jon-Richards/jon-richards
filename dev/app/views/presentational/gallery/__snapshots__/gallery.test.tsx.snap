// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The gallery component. Should match the snapshot. 1`] = `
.emotion-0 {
  display: grid;
  width: 100%;
  height: auto;
  margin-bottom: 2rem;
  grid-template-columns: 1fr;
  grid-template-rows: repeat(auto-fill,auto);
  -webkit-column-gap: 2rem;
  column-gap: 2rem;
  row-gap: 2rem;
}

@media (min-width:720px) {
  .emotion-0 {
    grid-template-columns: 1fr 1fr;
  }
}

@media (min-width:1440px) {
  .emotion-0 {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

<div
  class="emotion-0"
/>
`;
